cmd/kubeadm/app/images/interface.go

```diff
@@ -0,0 +1,107 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package images
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	kubeadmapiv1alpha2 "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha2"
+	utilsexec "k8s.io/utils/exec"
+)
+
+const (
+	// ImageNotFoundErrorMessage is the message crictl returns when an image is not found with `inspecti`
+	ImageNotFoundErrorMessage = "no such image"
+)
+
+// Puller is an interface for pulling images
+type Puller interface {
+	Pull(string) error
+}
+
+// Existence is an interface to determine if an image exists on the system
+type Existence interface {
+	Exists(string) (int, error)
+}
+
+// Images defines the set of behaviors needed for images realting to the CRI
+type Images interface {
+	Puller
+	Existence
+}
+
+// CRInterfacer is a struct that interfaces with the container runtime
+type CRInterfacer struct {
+	criSocket  string
+	exec       utilsexec.Interface
+	crictlPath string
+}
+
+// NewCRInterfacer sets up and returns a CRInterfacer
+func NewCRInterfacer(execer utilsexec.Interface, criSocket string) (*CRInterfacer, error) {
+	crictlPath, err := execer.LookPath("crictl")
+	if err != nil && criSocket != kubeadmapiv1alpha2.DefaultCRISocket {
+		return nil, fmt.Errorf("crictl is required for non docker container runtimes: %v", err)
+	}
+	return &CRInterfacer{
+		exec:       execer,
+		criSocket:  criSocket,
+		crictlPath: crictlPath,
+	}, nil
+}
+
+// Pull pulls the actual image using either crictl or docker
+func (cri *CRInterfacer) Pull(image string) error {
+	if cri.criSocket != kubeadmapiv1alpha2.DefaultCRISocket {
+		return cri.exec.Command(cri.crictlPath, "-r", cri.criSocket, "pull", image).Run()
+	}
+	return cri.exec.Command("sh", "-c", fmt.Sprintf("docker pull %v", image)).Run()
+}
+
+// Exists checks to see if the image exists on the system already
+// Returns
+//    * -1 if an error occurred with the exec command
+//    * 0  if no error occurred but image is not found
+//    * 1  if no error occurred and the image is found
+func (cri *CRInterfacer) Exists(image string) (int, error) {
+	if cri.criSocket != kubeadmapiv1alpha2.DefaultCRISocket {
+		err := cri.exec.Command(cri.crictlPath, "-r", cri.criSocket, "inspecti", image).Run()
+		if err == utilsexec.ErrExecutableNotFound {
+			return -1, err
+		}
+		if err != nil {
+			if strings.Contains(err.Error(), ImageNotFoundErrorMessage) {
+				return 0, nil
+			}
+			return -1, err
+		}
+		return 1, nil
+	}
+	var buf bytes.Buffer
+	// docker images *always* returns exit code 0 even if the image doesn't exist so have to check the output
+	dockerCmd := cri.exec.Command("sh", "-c", fmt.Sprintf("docker images %v", image))
```

Comment:
We should lookup the `dockerPath` and add it to `CRInterfacer`.

This will allow us to avoid an unecessary subshell.

Additionally, we can add
`else if cri.dockerPath != "" {`

The default case would be to return `ImageCheckerUnavailable`(-2) when neither `crictl` or `docker` are available to check for images.
This will allow us to return a user-friendly error message to people who could potentially be using containerd or rkt without crictl in their PATH.
The error message would be a better signal for users to skip the pre-flight check than "`sh: 1: docker: not found`"